src/or/proto_socks.c

```diff
@@ -719,13 +716,26 @@ fetch_from_buf_socks(buf_t *buf, socks_request_t *req,
     tor_assert(head && datalen >= 2);
     want_length = 0;
 
-    res = parse_socks(head, datalen, req, log_sockstype,
-                      safe_socks, &n_drain, &want_length);
+    socks_res = parse_socks(head, datalen, req, log_sockstype,
+                            safe_socks, &n_drain, &want_length);
 
-    if (res == -1)
+    if (socks_res == SOCKS_RESULT_INVALID)
       buf_clear(buf);
-    else if (n_drain > 0)
+    else if (socks_res != SOCKS_RESULT_TRUNCATED && n_drain > 0)
       buf_drain(buf, n_drain);
+
+    switch (socks_res) {
+      case SOCKS_RESULT_INVALID:
+        res = -1;
+        break;
+      case SOCKS_RESULT_DONE:
+        res = 1;
+        break;
+      case SOCKS_RESULT_TRUNCATED:
```

Comment:
This should have a "falls through" comment so GCC doesn't complain.

Reply:
Fixed in 90f61531dc9880cbbc1ec2c07816090abd7bfaf3.